File: API-Practice/Random.Api.ts
// Source: https://abhiappmobiledeveloper.medium.com/free-api-huge-list-of-public-apis-for-developertesting-b9cf371282b3

type APIResponse = {
    "ID Nation": string;
    Nation: string;
    "ID Year": number;
    Year: number;
    Population: number;
    "Slug Nation": string;
};

async function practiceRESTAPIForCodingAssessments(apiUrl: string) {
    try {
        const response = await fetch(apiUrl, {
            headers: {
                "Content-Type": "application/json",
                "Accept": "application/json",
            },
        });

        if (!response.ok || !response.status.toString().startsWith("2")) {
            throw new Error(`Response status: ${response.status}`);
        }

        const apiResponse = (await response.json() as { data: Readonly<APIResponse[]> }).data;
        console.log(JSON.stringify(apiResponse, null, 2));
        console.log(apiResponse.filter((item) => item["ID Year"] === 2018).length);
    }
    catch (err) {
        const error = err as Error;
        console.log("Error: ", error.message);
    }
}

const apiUrls = [
    "https://datausa.io/api/data?drilldowns=Nation&measures=Population"
];

practiceRESTAPIForCodingAssessments(apiUrls[0]);

File: API-Practice/random_api.py
# Source: https://abhiappmobiledeveloper.medium.com/free-api-huge-list-of-public-apis-for-developertesting-b9cf371282b3

import requests
import json

class Result:
    ID_nation: str
    nation: str
    ID_year: int
    year: int
    population: int
    slug_nation: str
    
    def __init__(self, ID_nation: str, nation: str, ID_year: int, year: int, population: int, slug_nation: str) -> None:
        self.ID_nation = ID_nation
        self.nation = nation
        self.ID_year = ID_year
        self.year = year
        self.population = population
        self.slug_nation = slug_nation

def main() -> None:
    API_URL = "http://universities.hipolabs.com/search?country=United+States"
    response = requests.get(API_URL, headers={
        "Accept": "application/json",
        "Content-Type": "application/json",
    })
    
    if not str(response.status_code).startswith("2"):
        print(f"Error: {response.status_code}")
        exit(1)  
    
    res_body = response.json()
    print(json.dumps(res_body, indent=4))
    
    print(len(res_body))
    data = res_body
    
    # data = res_body.get("data")
    # print(len(data))
    
    results: list[Result] = []
    for _, v in enumerate(data):
        result = Result(
            ID_nation=v.get("ID nation"),
            nation=v.get("nation"),
            ID_year=v.get("ID year"),
            year=v.get("year"),
            population=v.get("Population"),
            slug_nation=v.get("Slug nation"),
        )
        results.append(result)
    print(results)
    return None

if __name__ == "__main__":
    main()